File: sast-python2/case/completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_001_F.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = false
# evaluation item = 完整度->单应用跟踪完整度->函数和方法调用->高阶函数
# scene introduction = 一阶
# level = 2
# bind_url = completeness/single_app_tracing/function_call/higher_order_function/higher_order_function_001_F
# evaluation information end
import os


def higher_order_function_001_F(taint_src):
    def f(a, b):
        c = u'_'

        def inner():
            return a + b + c

        return inner

    taint_sink(f(u'aa', u'_')())


def taint_sink(o):
    os.system(o)


if __name__ == u'__main__':
    taint_src = u"taint_src_value"
    higher_order_function_001_F(taint_src)

File: sast-php/case/completeness/single_app_tracing/function_call/higher_order_function_001_F.php
<?php
// evaluation information start
// real case = false
// evaluation item = 完整度->单应用追踪->函数调用
// scene introduction = array_map 回调返回安全值
// level = 2
// bind_url = completeness/single_app_tracing/function_call/higher_order_function_001_F
// evaluation information end

function higher_order_function_001_F($__taint_src) {
    $arr = [$__taint_src];
    $mapped = array_map(function($item) {
        return "safe_mapped";
    }, $arr);
    __taint_sink($mapped[0]);
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
higher_order_function_001_F($taint_src);
